From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byteflux <byte@byteflux.net>
Date: Wed, 2 Mar 2016 12:27:07 -0600
Subject: [PATCH] Configurable lava flow speed


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index dbd82d5a9e3f3954df5ab913d94581fa1e298c9a..b0b3033e789a5900ab9e2732bbe33a0afde99d53 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -190,4 +190,11 @@ public class PaperWorldConfig {
         fastDrainLava = getBoolean("fast-drain.lava", false);
         fastDrainWater = getBoolean("fast-drain.water", false);
     }
+
+    public int lavaFlowSpeedNormal;
+    public int lavaFlowSpeedNether;
+    private void lavaFlowSpeeds() {
+        lavaFlowSpeedNormal = getInt("lava-flow-speed.normal", 30);
+        lavaFlowSpeedNether = getInt("lava-flow-speed.nether", 10);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/BlockFlowing.java b/src/main/java/net/minecraft/server/BlockFlowing.java
index 3aaa19b2fee4c7db5d89baf119c9e28036c9b64e..739b9aac3642747621efd08d63223fd2a70b759e 100644
--- a/src/main/java/net/minecraft/server/BlockFlowing.java
+++ b/src/main/java/net/minecraft/server/BlockFlowing.java
@@ -272,6 +272,9 @@ public class BlockFlowing extends BlockFluids {
      * Paper - Get flow speed. Throttle if its water and flowing adjacent to lava
      */
     public int getFlowSpeed(World world, BlockPosition blockposition) {
+        if (this.material == Material.LAVA) {
+            return world.worldProvider.isSkyMissing() ? world.paperConfig.lavaFlowSpeedNether : world.paperConfig.lavaFlowSpeedNormal;
+        }
         if (this.material == Material.WATER && (
                 world.getType(blockposition.north(1)).getBlock().material == Material.LAVA ||
                         world.getType(blockposition.south(1)).getBlock().material == Material.LAVA ||
diff --git a/src/main/java/net/minecraft/server/WorldProvider.java b/src/main/java/net/minecraft/server/WorldProvider.java
index f3eb2a797f6cc036fed7ada4522640c3f8f5a2e7..d0265f960f15aa84a501fb3cf4d3f56a35bc3089 100644
--- a/src/main/java/net/minecraft/server/WorldProvider.java
+++ b/src/main/java/net/minecraft/server/WorldProvider.java
@@ -114,6 +114,7 @@ public abstract class WorldProvider {
         return this.f;
     }
 
+    public final boolean isSkyMissing() { return this.n(); } // Paper - OBFHELPER
     public boolean n() {
         return this.e;
     }
