From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Fri, 16 Dec 2016 16:03:19 -0600
Subject: [PATCH] Don't let fishinghooks use portals


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 2677cb68dc3b2e7749c1d9d7a1ec348e87816f2c..feee50227ef70c0701f353c13de18d686036fd47 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -157,7 +157,7 @@ public abstract class Entity implements ICommandListener, KeyedObject { // Paper
     public boolean ah;
     public boolean impulse;
     public int portalCooldown;
-    protected boolean ak;
+    protected boolean ak; public boolean inPortal() { return ak; } // Paper - OBFHELPER
     protected int al;
     public int dimension;
     protected BlockPosition an;
diff --git a/src/main/java/net/minecraft/server/EntityFishingHook.java b/src/main/java/net/minecraft/server/EntityFishingHook.java
index 339d1f1b1445291a420b8373cc72b2a079130319..7f011aef80e02d65608d1d495c98103f83232272 100644
--- a/src/main/java/net/minecraft/server/EntityFishingHook.java
+++ b/src/main/java/net/minecraft/server/EntityFishingHook.java
@@ -185,6 +185,12 @@ public class EntityFishingHook extends Entity {
             this.motY *= 0.92D;
             this.motZ *= 0.92D;
             this.setPosition(this.locX, this.locY, this.locZ);
+
+            // Paper start - These shouldn't be going through portals
+            if (this.inPortal()) {
+                this.die();
+            }
+            // Paper end
         }
     }
 
