From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 23 Jul 2018 22:18:31 -0400
Subject: [PATCH] Mark chunk dirty anytime entities change to guarantee it
 saves


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 71d861753d9d799c5fe44abd7ece64eaa2227d19..e5c900a85f9e5e98d2a15adfa8c45fc21d877af3 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -683,6 +683,7 @@ public class Chunk {
         entity.ad = this.locZ;
         this.entitySlices[k].add(entity);
         // Paper start
+        this.markDirty();
         entity.setCurrentChunk(this);
         entityCounts.increment(entity.getMinecraftKeyString());
         if (entity instanceof EntityItem) {
@@ -725,6 +726,7 @@ public class Chunk {
 
         // Paper start
         if (!this.entitySlices[i].remove(entity)) { return; }
+        this.markDirty();
         entity.setCurrentChunk(null);
         entityCounts.decrement(entity.getMinecraftKeyString());
         if (entity instanceof EntityItem) {
