From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: kashike <kashike@vq.lc>
Date: Wed, 15 Aug 2018 01:26:09 -0700
Subject: [PATCH] Allow disabling armour stand ticking


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index c3bd82692d7a47d6af73822833cfc58025775daa..ed1475351221434aaca9996def0b57c586add849 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -584,4 +584,10 @@ public class PaperWorldConfig {
                 break;
         }
     }
+
+    public boolean armorStandTick = true;
+    private void armorStandTick() {
+        this.armorStandTick = this.getBoolean("armor-stands-tick", this.armorStandTick);
+        log("ArmorStand ticking is " + (this.armorStandTick ? "enabled" : "disabled") + " by default");
+    }
 }
diff --git a/src/main/java/net/minecraft/server/EntityArmorStand.java b/src/main/java/net/minecraft/server/EntityArmorStand.java
index df0d66ad0c04944de05bf37e71db999522f04fb1..c72158882648f55d3268383a1a6f071702cea2c8 100644
--- a/src/main/java/net/minecraft/server/EntityArmorStand.java
+++ b/src/main/java/net/minecraft/server/EntityArmorStand.java
@@ -51,6 +51,7 @@ public class EntityArmorStand extends EntityLiving {
     public Vector3f leftLegPose;
     public Vector3f rightLegPose;
     public boolean canMove = true; // Paper
+    public boolean canTick = true; // Paper - armour stand ticking
 
     public EntityArmorStand(World world) {
         super(world);
@@ -64,6 +65,7 @@ public class EntityArmorStand extends EntityLiving {
         this.rightLegPose = EntityArmorStand.bw;
         this.noclip = this.isNoGravity();
         this.setSize(0.5F, 1.975F);
+        if (world != null) this.canTick = world.paperConfig.armorStandTick; // Paper - armour stand ticking
     }
 
     public EntityArmorStand(World world, double d0, double d1, double d2) {
@@ -211,6 +213,7 @@ public class EntityArmorStand extends EntityLiving {
         }
 
         nbttagcompound.set("Pose", this.C());
+        nbttagcompound.setBoolean("Paper.CanTick", this.canTick); // Paper - persist no tick setting
     }
 
     public void a(NBTTagCompound nbttagcompound) {
@@ -242,6 +245,10 @@ public class EntityArmorStand extends EntityLiving {
         this.setMarker(nbttagcompound.getBoolean("Marker"));
         this.bC = !this.isMarker();
         this.noclip = this.isNoGravity();
+        // Paper - persist no tick
+        if (nbttagcompound.hasKey("Paper.CanTick")) {
+            this.canTick = nbttagcompound.getBoolean("Paper.CanTick");
+        }
         NBTTagCompound nbttagcompound1 = nbttagcompound.getCompound("Pose");
 
         this.g(nbttagcompound1);
@@ -568,6 +575,7 @@ public class EntityArmorStand extends EntityLiving {
     }
 
     public void B_() {
+        if (!this.canTick) return;// Paper
         super.B_();
         Vector3f vector3f = (Vector3f) this.datawatcher.get(EntityArmorStand.b);
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftArmorStand.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftArmorStand.java
index 8a06cb1656e2ba396d6abcd0f0fdb51939f6407b..91b7bc2edd6e36c6a782872595b869a1c9c87636 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftArmorStand.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftArmorStand.java
@@ -221,4 +221,16 @@ public class CraftArmorStand extends CraftLivingEntity implements ArmorStand {
     public void setCanMove(boolean move) {
         getHandle().canMove = move;
     }
+
+    // Paper start
+    @Override
+    public boolean canTick() {
+        return this.getHandle().canTick;
+    }
+
+    @Override
+    public void setCanTick(final boolean tick) {
+        this.getHandle().canTick = tick;
+    }
+    // Paper end
 }
