From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 12 Oct 2018 22:41:29 -0400
Subject: [PATCH] MC-54026: Backport 1.13 client desync fix on fast tools


diff --git a/src/main/java/net/minecraft/server/PlayerInteractManager.java b/src/main/java/net/minecraft/server/PlayerInteractManager.java
index fcb64666e3929fdffa62fa0d1da7bdbf3c4272f7..3d86798248d2256f394d1f4f848ed5a2f2f0695c 100644
--- a/src/main/java/net/minecraft/server/PlayerInteractManager.java
+++ b/src/main/java/net/minecraft/server/PlayerInteractManager.java
@@ -197,6 +197,7 @@ public class PlayerInteractManager {
                 int i = (int) (f * 10.0F);
 
                 this.world.c(this.player.getId(), blockposition, i);
+                this.player.playerConnection.sendPacket(new PacketPlayOutBlockChange(world, blockposition)); // Paper - MC-54026 - backport from 1.13
                 this.k = i;
             }
 
