From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach@zachbr.io>
Date: Sun, 7 Apr 2019 06:22:54 -0400
Subject: [PATCH] Allow disabling village sieges


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index ba299afc40bdeadbdc8938495dba7c3572528e96..319b5b95db6393be4a89a245dd2722394a806510 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -587,4 +587,10 @@ public class PaperWorldConfig {
         this.armorStandTick = this.getBoolean("armor-stands-tick", this.armorStandTick);
         log("ArmorStand ticking is " + (this.armorStandTick ? "enabled" : "disabled") + " by default");
     }
+
+    public boolean villageSiegesEnabled = true;
+    private void villageSiegesEnabled() {
+        this.villageSiegesEnabled = getBoolean("game-mechanics.village-sieges-enabled", this.villageSiegesEnabled);
+        log("Village sieges are " + (this.villageSiegesEnabled ? "enabled" : "disabled"));
+    }
 }
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index d29420dd49da191ee2a4de1a18377fc3da969a66..e545e9764a4aa117896ca1b3e55d2613642d7244 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -334,7 +334,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
         this.methodProfiler.c("village");
         timings.doVillages.startTiming(); // Spigot
         this.villages.tick();
-        this.siegeManager.a();
+        if (paperConfig.villageSiegesEnabled) { this.siegeManager.a(); } // Paper - Allow disabling village sieges
         timings.doVillages.stopTiming(); // Spigot
         this.methodProfiler.c("portalForcer");
         timings.doPortalForcer.startTiming(); // Spigot
