From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <ilyy@outlook.jp>
Date: Mon, 11 Jan 2021 21:13:40 +0900
Subject: [PATCH] Disable async safety


diff --git a/src/main/java/org/spigotmc/AsyncCatcher.java b/src/main/java/org/spigotmc/AsyncCatcher.java
index e44c2301654b8ce8e7a1135ac739bc267048a967..11b5c46283eca67550b6ecf01e21c04536530732 100644
--- a/src/main/java/org/spigotmc/AsyncCatcher.java
+++ b/src/main/java/org/spigotmc/AsyncCatcher.java
@@ -10,7 +10,7 @@ public class AsyncCatcher
 
     public static void catchOp(String reason)
     {
-        if ( enabled && Thread.currentThread() != MinecraftServer.getServer().primaryThread )
+        if (false && enabled && Thread.currentThread() != MinecraftServer.getServer().primaryThread ) // 2p2t - disable safety
         {
             throw new IllegalStateException( "Asynchronous " + reason + "!" );
         }
