From 35a4d0ceaeef16dddad6a7df85d4d3d84e681b2e Mon Sep 17 00:00:00 2001
From: acrylic-style <ilyy@outlook.jp>
Date: Tue, 12 Jan 2021 01:32:17 +0900
Subject: [PATCH] Drop support for old events


diff --git a/src/main/java/net/minecraft/server/LoginListener.java b/src/main/java/net/minecraft/server/LoginListener.java
index be43d929..53fa4399 100644
--- a/src/main/java/net/minecraft/server/LoginListener.java
+++ b/src/main/java/net/minecraft/server/LoginListener.java
@@ -303,7 +303,7 @@ public class LoginListener implements PacketLoginInListener, ITickable {
                             uniqueId = i.getId();
                             // Paper end
 
-                            if (PlayerPreLoginEvent.getHandlerList().getRegisteredListeners().length != 0) {
+                            if (false && PlayerPreLoginEvent.getHandlerList().getRegisteredListeners().length != 0) { // 2p2t - drop support for evil plugins
                                 final PlayerPreLoginEvent event = new PlayerPreLoginEvent(playerName, address, uniqueId);
                                 if (asyncEvent.getResult() != PlayerPreLoginEvent.Result.ALLOWED) {
                                     event.disallow(asyncEvent.getResult(), asyncEvent.getKickMessage());
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 59d52632..d268be30 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1246,7 +1246,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
         }
 
         boolean isSync = packetplayinchat.a().startsWith("/");
-        if (packetplayinchat.a().startsWith("/")) {
+        if (isSync) { // 2p2t
             PlayerConnectionUtils.ensureMainThread(packetplayinchat, this, this.player.x());
         }
         // CraftBukkit end
@@ -1406,7 +1406,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
             AsyncPlayerChatEvent event = new AsyncPlayerChatEvent(async, player, s, new LazyPlayerSet(minecraftServer));
             this.server.getPluginManager().callEvent(event);
 
-            if (PlayerChatEvent.getHandlerList().getRegisteredListeners().length != 0) {
+            if (false && PlayerChatEvent.getHandlerList().getRegisteredListeners().length != 0) { // 2p2t - disable, drop support for evil plugins
                 // Evil plugins still listening to deprecated event
                 final PlayerChatEvent queueEvent = new PlayerChatEvent(player, event.getMessage(), event.getFormat(), event.getRecipients());
                 queueEvent.setCancelled(event.isCancelled());
-- 
2.21.0.windows.1

