From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <ilyy@outlook.jp>
Date: Mon, 18 Jan 2021 06:06:01 +0900
Subject: [PATCH] don't break blocks from wither itself


diff --git a/src/main/java/net/minecraft/server/EntityWither.java b/src/main/java/net/minecraft/server/EntityWither.java
index 9678526059621cd31a9a0dccc8ff803d80e9082a..8d26b1356d13a9920862158c3cef40ff518c718c 100644
--- a/src/main/java/net/minecraft/server/EntityWither.java
+++ b/src/main/java/net/minecraft/server/EntityWither.java
@@ -330,7 +330,7 @@ public class EntityWither extends EntityMonster implements IRangedEntity {
                                         continue;
                                     }
                                     // CraftBukkit end
-                                    flag = this.world.setAir(blockposition, true) || flag;
+                                    flag = this.world.setAir(blockposition, false) || flag;
                                 }
                             }
                         }
