From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 4 Jul 2018 15:29:21 -0400
Subject: [PATCH] Vex#getOwner API

Get's the NPC that summoned this Vex

diff --git a/src/main/java/org/bukkit/entity/Vex.java b/src/main/java/org/bukkit/entity/Vex.java
index a2f2fcaec8d3ddc4b2837ce69912ed42a042dddf..540135c3503a61a7b00f671573300b89d9d927b6 100644
--- a/src/main/java/org/bukkit/entity/Vex.java
+++ b/src/main/java/org/bukkit/entity/Vex.java
@@ -1,6 +1,13 @@
 package org.bukkit.entity;
 
+import com.destroystokyo.paper.entity.SentientNPC;
+
 /**
  * Represents a Vex.
  */
-public interface Vex extends Monster { }
+public interface Vex extends Monster {
+    /**
+     * @return What Entity (most likely an Evoker, but not guaranteed) summoned this Vex
+     */
+    SentientNPC getOwner(); // Paper
+}
